src/oatpp/core/provider/Pool.hpp

```diff
@@ -189,100 +190,91 @@ class PoolTemplate : public oatpp::base::Countable, public async::CoroutineWaitL
+  PoolTemplate(const std::shared_ptr<Provider<TResource>>& provider, v_int64 maxResources, v_int64 maxResourceTTL, const std::chrono::duration<v_int64, std::micro>& timeout)
     : m_provider(provider)
-    , m_counter(0)
     , m_maxResources(maxResources)
     , m_maxResourceTTL(maxResourceTTL)
-    , m_running(true)
-    , m_finished(false)
+    , m_timeout(timeout)
   {}
 
-  void startCleanupTask(const std::shared_ptr<PoolTemplate>& _this) {
+  static void startCleanupTask(const std::shared_ptr<PoolTemplate>& _this) {
     std::thread poolCleanupTask(cleanupTask, _this);
     poolCleanupTask.detach();
   }
 
-public:
-
-  static std::shared_ptr<PoolTemplate> createShared(const std::shared_ptr<Provider<TResource>>& provider,
-                                                    v_int64 maxResources,
-                                                    const std::chrono::duration<v_int64, std::micro>& maxResourceTTL)
-  {
-    /* "new" is called directly to keep constructor private */
-    auto ptr = std::shared_ptr<PoolTemplate>(new PoolTemplate(provider, maxResources, maxResourceTTL.count()));
-    ptr->startCleanupTask();
-    return ptr;
-  }
-
-  virtual ~PoolTemplate() {
-    stop();
-  }
+  static std::shared_ptr<TResource> get(const std::shared_ptr<PoolTemplate>& _this) {
+    const auto waitUntil = _this->m_timeout == std::chrono::microseconds::zero()
+                             ? std::chrono::steady_clock::time_point::max()
+                             : std::chrono::steady_clock::now() + _this->m_timeout;
+    auto readyPredicate = [&_this]() { return !_this->m_running || !_this->m_bench.empty() || _this->m_counter < _this->m_maxResources; };
 
-  std::shared_ptr<TResource> get(const std::shared_ptr<PoolTemplate>& _this) {
-
-    {
-
-      std::unique_lock<std::mutex> guard(m_lock);
-
-      while (m_running && m_bench.size() == 0 && m_counter >= m_maxResources ) {
-        m_condition.wait(guard);
-      }
-
-      if(!m_running) {
-        return nullptr;
-      }
+    std::unique_lock<std::mutex> guard{_this->m_lock};
+    if (!_this->m_condition.wait_until(guard, waitUntil, std::move(readyPredicate))) {
+      return nullptr;
+    }
 
-      if (m_bench.size() > 0) {
-        auto record = m_bench.front();
-        m_bench.pop_front();
-        return std::make_shared<AcquisitionProxyImpl>(record.resource, _this);
-      } else {
-        ++ m_counter;
-      }
+    if(!_this->m_running) {
+      return nullptr;
+    }
 
+    if (_this->m_bench.size() > 0) {
+      auto record = _this->m_bench.front();
+      _this->m_bench.pop_front();
+      return std::make_shared<AcquisitionProxyImpl>(record.resource, _this);
+    } else {
+      ++ _this->m_counter;
     }
 
+    guard.unlock();
+
     try {
-      return std::make_shared<AcquisitionProxyImpl>(m_provider->get(), _this);
+      return std::make_shared<AcquisitionProxyImpl>(_this->m_provider->get(), _this);
```

Comment:
I was thinking: Since `_this->m_provider->get()` may block, don't we need a timeout here as well?

Reply:
As the original issue states - we want to implement a timeout for pool resource acquisition.
When there are no free resources available in the pool we want to be able to limit our waiting time.

On one hand, at this particular point, the resource is not yet allocated but on the other, our request for the resource is already being processed. So it's kinda a gray zone.

I propose to treat pool timeout as a "pool resource acquisition timeout". So there is no need for the provider's `get()` timeout since the pool is already allocated a slot for us.

Also, if we introduce timeout in `providers->get()` again - we'll return to where we began from:)